Add small pointer guide in 'Interfacing with External C Code' chapter (#3833)
cython/cython: docs/src/userguide/external_C_code.rst

Note that in all the cases below, you refer to the type in Cython code simply
as :c:type:`Foo`, not ``struct Foo``.

Pointers
--------
When interacting with a C-api there may be functions that require pointers as arguments.
Pointers are variables that contain a memory address to another variable.

For example::

    cdef extern from "<my_lib.h>":
        cdef void increase_by_one(int *my_var)

This function requires a pointer to an integer. In order to get the address from an existing

> variable use the ``&`` sign::

Reviewer: ```suggestion
variable use the ``&`` operator::
```

> This function requires a pointer to an integer. In order to get the address from an existing
> variable use the ``&`` operator::

Reviewer: ```suggestion
This function requires a pointer to an integer. Knowing the address of the integer allows
the function to modify the value in place, thus passing the changes back to its caller.
In order to get the address from an existing variable, use the ``&`` operator::
```

    cdef int some_int = 42

>     cdef int * some_int_pointer = &some_int

Reviewer: ```suggestion
    cdef int *some_int_pointer = &some_int
```

    increase_by_one(some_int_pointer)
    # Or without creating the extra variable
    increase_by_one(&some_int)

> If you want to manipulate the variable the pointer points to, you can access pointer by

Reviewer: ```suggestion
If you want to manipulate the variable the pointer points to, you can access it by
```

referencing its first element like you would in python ``my_pointer[0]``. For example::

    cdef void increase_by_one(int *my_var):
        my_var[0] += 1

> For a pointer 101 you can read `this tutorial at tutorialspoint

Reviewer: ```suggestion
For a deeper introduction to pointers, you can read `this tutorial at tutorialspoint
```

<https://www.tutorialspoint.com/cprogramming/c_pointers.htm>`_. For differences between

> Cython and C syntax for manipulating pointers checkout :ref:`statements_and_expressions`.

Reviewer: ```suggestion
Cython and C syntax for manipulating pointers, see :ref:`statements_and_expressions`.
```

Note that in all the cases below, you refer to the type in Cython code simply
as :c:type:`Foo`, not ``struct Foo``.

Pointers
--------
When interacting with a C-api there may be functions that require pointers as arguments.
Pointers are variables that contain a memory address to another variable.

For example::

    cdef extern from "<my_lib.h>":
        cdef void increase_by_one(int *my_var)

This function takes a pointer to an integer as argument.  Knowing the address of the
integer allows the function to modify the value in place, so that the caller can see
the changes afterwards.  In order to get the address from an existing variable,
use the ``&`` operator::

    cdef int some_int = 42
    cdef int *some_int_pointer = &some_int
    increase_by_one(some_int_pointer)
    # Or without creating the extra variable

>     increase_by_one(&some_int)

Reviewer: ```suggestion
    increase_by_one(&some_int)
    print(some_int)  # prints 44 (== 42+1+1)
```

Note that in all the cases below, you refer to the type in Cython code simply
as :c:type:`Foo`, not ``struct Foo``.

Pointers
--------
When interacting with a C-api there may be functions that require pointers as arguments.
Pointers are variables that contain a memory address to another variable.

For example::

    cdef extern from "<my_lib.h>":
        cdef void increase_by_one(int *my_var)

This function takes a pointer to an integer as argument.  Knowing the address of the
integer allows the function to modify the value in place, so that the caller can see
the changes afterwards.  In order to get the address from an existing variable,
use the ``&`` operator::

    cdef int some_int = 42
    cdef int *some_int_pointer = &some_int
    increase_by_one(some_int_pointer)
    # Or without creating the extra variable
    increase_by_one(&some_int)
    print(some_int)  # prints 44 (== 42+1+1)

If you want to manipulate the variable the pointer points to, you can access it by

> referencing its first element like you would in python ``my_pointer[0]``. For example::

Reviewer 2: maybe worth mentioning the operator dereference as well? Coming from C that would be the more natural way of doing it instead of abusing the array notation.

Reviewer: OTOH, coming from C, you would probably know what a pointer is and would rather skip over this section completely. Or need entirely different information that explains the different meaning of `*x` in Python and,thus, Cython. That's a different PR, I think.
